jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -0,0 +1,259 @@
+
+import org.eclipse.jetty.http.BadMessageException;
+import org.eclipse.jetty.http.HttpStatus;
+
+class AsyncContentProducer implements ContentProducer
+{
+    private final HttpChannel _httpChannel;
+    private HttpInput.Interceptor _interceptor;
+    private HttpInput.Content _rawContent;
+    private HttpInput.Content _transformedContent;
+    private boolean _error;
+
+    private long _firstByteTimeStamp = Long.MIN_VALUE;
+    private long _rawContentArrived;
+
+    public AsyncContentProducer(HttpChannel httpChannel)
+    {
+        _httpChannel = httpChannel;
+    }
+
+    @Override
+    public void recycle()
+    {
+        _interceptor = null;
+        _rawContent = null;
+        _transformedContent = null;
+        _error = false;
+        _firstByteTimeStamp = Long.MIN_VALUE;
+        _rawContentArrived = 0L;
+    }
+
+    @Override
+    public HttpInput.Interceptor getInterceptor()
+    {
+        return _interceptor;
+    }
+
+    @Override
+    public void setInterceptor(HttpInput.Interceptor interceptor)
+    {
+        this._interceptor = interceptor;
+    }
+
+    @Override
+    public int available()
+    {
+        HttpInput.Content content = nextTransformedContent();
+        return content == null ? 0 : content.remaining();
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        return _rawContent != null;
+    }
+
+    @Override
+    public boolean isError()
+    {
+        return _error;
+    }
+
+    @Override
+    public void checkMinDataRate()
+    {
+        long minRequestDataRate = _httpChannel.getHttpConfiguration().getMinRequestDataRate();
+        if (minRequestDataRate > 0 && _firstByteTimeStamp != Long.MIN_VALUE)
+        {
+            long period = System.nanoTime() - _firstByteTimeStamp;
+            if (period > 0)
+            {
+                long minimumData = minRequestDataRate * TimeUnit.NANOSECONDS.toMillis(period) / TimeUnit.SECONDS.toMillis(1);
+                if (getRawContentArrived() < minimumData)
+                {
+                    BadMessageException bad = new BadMessageException(HttpStatus.REQUEST_TIMEOUT_408,
+                        String.format("Request content data rate < %d B/s", minRequestDataRate));
+                    if (_httpChannel.getState().isResponseCommitted())
+                        _httpChannel.abort(bad);
+                    failCurrentContent(bad);
+                    throw bad;
+                }
+            }
+        }
+    }
+
+    @Override
+    public long getRawContentArrived()
+    {
+        return _rawContentArrived;
+    }
+
+    @Override
+    public boolean consumeAll(Throwable x)
+    {
+        failCurrentContent(x);
+        // A specific HttpChannel mechanism must be used as the normal produceContent()
+        // mechanism does not guarantee that all available content will be returned;
+        // tests in StreamResetTest cover this.
+        _httpChannel.failContent(x);
+        HttpInput.Content c = produceRawContent();
```

Comment:
`consumeAll()` needs to return true if EOF was reached, so calling `httpChannel.nextContent()` then checking if the returned content is EOF does the trick since once you reach EOF, you can continuously call `httpChannel.nextContent()` and always be served a EofContent.

Reply:
Indeed. This can be solved by making `failContent(x)` return a boolean to indicate if EOF was reached. Let me do this.

Reply:
Done.